controllers/resources/authconfig.go

```diff
@@ -0,0 +1,210 @@
+		TypeMeta: v1.TypeMeta{
+			APIVersion: "authorino.kuadrant.io/v1beta2",
+			Kind:       "AuthConfig",
+		},
+	}
+
+	err := c.client.Get(ctx, key, authConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not GET authconfig %s", key)
+	}
+	return authConfig, nil
+}
+
+func (c *clientAuthConfigStore) Remove(ctx context.Context, key types.NamespacedName) error {
+	authConfig := authorinov1beta2.AuthConfig{}
+	authConfig.Name = key.Name
+	authConfig.Namespace = key.Namespace
+	return errors.Wrapf(c.client.Delete(ctx, &authConfig), "could not DELETE authconfig %s", key)
+}
+
+func (c *clientAuthConfigStore) Create(ctx context.Context, authConfig *authorinov1beta2.AuthConfig) error {
+	return errors.Wrapf(c.client.Create(ctx, authConfig), "could not CREATE authconfig %s/%s", authConfig.Namespace, authConfig.Name)
+}
+
+func (c *clientAuthConfigStore) Update(ctx context.Context, authConfig *authorinov1beta2.AuthConfig) error {
+	return errors.Wrapf(c.client.Update(ctx, authConfig), "could not UPDATE authconfig %s/%s", authConfig.Namespace, authConfig.Name)
+}
+
+type kserveAuthTypeDetector struct {
+	client client.Client
+}
+
+func NewKServeAuthTypeDetector(client client.Client) AuthTypeDetector {
+	return &kserveAuthTypeDetector{
+		client: client,
+	}
+}
+
+func (k *kserveAuthTypeDetector) Detect(ctx context.Context, isvc *kservev1beta1.InferenceService) (AuthType, error) {
+	if strings.ToLower(isvc.Annotations["enable-auth"]) == "true" {
+		return UserDefined, nil
+	}
+	return Anonymous, nil
+}
+
+type kserveInferenceServiceHostExtractor struct {
+}
+
+func NewKServeInferenceServiceHostExtractor() InferenceServiceHostExtractor {
+	return &kserveInferenceServiceHostExtractor{}
+}
+
+/*
+in: caikit-example-isvc-kserve-demo.apps-crc.testing
+
+out:
+caikit-example-isvc.kserve-demo.svc.cluster.local
+caikit-example-isvc-kserve-demo.apps-crc.testing
+caikit-example-isvc-predictor-kserve-demo.apps-crc.testing
+caikit-example-isvc-predictor.kserve-demo
+caikit-example-isvc-predictor.kserve-demo.svc
+caikit-example-isvc-predictor.kserve-demo.svc.cluster.local
+*/
+func (k *kserveInferenceServiceHostExtractor) Extract(isvc *kservev1beta1.InferenceService) []string {
+	statusHost := isvc.Status.URL.Host
+
+	// short cut exit, something is wrong with the expected pattern
+	if !strings.Contains(statusHost, isvc.Namespace) {
+		return []string{statusHost}
+	}
+
+	hosts := []string{}
+
+	base := strings.ReplaceAll(statusHost[0:strings.Index(statusHost, isvc.Namespace)-1], "-predictor", "")
+	basePre := base + "-predictor"
+
+	// caikit-example-isvc-kserve-demo.apps-crc.testing
+	hosts = append(hosts, base+statusHost[strings.Index(statusHost, isvc.Namespace)-1:])
+	// // caikit-example-isvc.kserve-demo.svc.cluster.local
+	hosts = append(hosts, base+"."+isvc.Namespace+".svc.cluster.local")
+
+	// caikit-example-isvc-predictor-kserve-demo.apps-crc.testing
+	hosts = append(hosts, basePre+statusHost[strings.Index(statusHost, isvc.Namespace)-1:])
```

Comment:
There's only two of them, sadly. We could surly include those as well (in case we didn't infer them) as a safety. The reason for _all_ the hostnames is simply that that's what the VirtualServices setup. 

I guess we could read all the hosts from the VirtualServices instead. Then we "only" need to infer the VirtualService names.

Reply:
> There's only two of them, sadly.
So, yes, there are two. This is all what you would get in the current release.
On the upcoming release, we will have three because `status.url` will become `https://sklearn-v2-iris-kserve-test.apps-crc.testing` (without the `predictor`).

From these three, I'd think the remaining work is to infer the other shortened in-cluster hostnames.

> I guess we could read all the hosts from the VirtualServices instead.

We could... and the name of the `VirtualService` seems [to be the same name as the InferenceService](https://github.com/opendatahub-io/kserve/blob/release-v0.11.1/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go#L442).

The problem that I see around inferring the hostnames is that these does not necessarily come from a single VirtualService, and we may need to protect not only the predictor pod.

Try the Transformer example for a more complete list of hostnames that may need to be protected: https://kserve.github.io/website/latest/modelserving/v1beta1/transformer/feast/#create-the-inferenceservice.

I haven't looked at it but, AFAIK, the VS that KServe creates will switch from using `predictor` hostnames to using `transformer` hostnames. Yet, the ISVC will still contain all involved hostames on several `url` fields.

Reply:
https://github.com/opendatahub-io/odh-model-controller/pull/130/commits/7aefd5d604e7db684e462b6ef74305abaa9fc330